File: backend/domain/models/box.go
package models

import (
	"github.com/cu-3rd-party/cu-roadmap/backend/domain/enums"
	"github.com/google/uuid"
	"github.com/lib/pq"
	"gorm.io/gorm"
)

type Box struct {
	ID                   uuid.UUID              `gorm:"type:uuid;primaryKey"`
	Kind                 enums.BoxKind          `gorm:"type:varchar(20);not null;index"`
	Title                string                 `gorm:"type:varchar(255)"`
	CourseID             *uuid.UUID             `gorm:"type:uuid;index"`
	LogicalOp            *enums.LogicalOp       `gorm:"type:varchar(10)"`
	RequiredCount        int                    `gorm:"not null;default:0"`
	RequirementType      *enums.RequirementType `gorm:"type:varchar(20)"`
	Specializations      pq.StringArray         `gorm:"type:text[]"`
	AdmissionYear        *int                   `gorm:"index"`
	MajorTrack           *string                `gorm:"type:varchar(64)"`
	Course               *Course                `gorm:"foreignKey:CourseID"`
	OutgoingRequirements []BoxEdge              `gorm:"foreignKey:ParentBoxID"`
}

func (b *Box) BeforeCreate(tx *gorm.DB) error {
	if b.ID == uuid.Nil {
		b.ID = uuid.New()
	}
	return nil
}

File: src/Domain/Entities/Box.cs
namespace CuRoadmap.Domain.Entities;

public class Box
{
    public Guid Id { get; set; }
    public Enums.BoxKind Kind { get; set; }
    public string Title { get; set; } = string.Empty;
    public Guid? CourseId { get; set; }
    public Enums.LogicalOp? LogicalOp { get; set; }
    public int RequiredCount { get; set; }
    public Enums.RequirementType? RequirementType { get; set; }
    public string[] Specializations { get; set; } = [];
    public string[] MandatorySpecializations { get; set; } = [];
    public int? AdmissionYear { get; set; }
    public string? MajorTrack { get; set; }

    public Course? Course { get; set; }
    public ICollection<BoxEdge> OutgoingRequirements { get; set; } = new List<BoxEdge>();
    public ICollection<BoxEdge> IncomingRequirements { get; set; } = new List<BoxEdge>();
}
